File: Go/Lexer/lexer/lexer.go
package lexer

import (
	token "golex/token"
	"unicode"
)

type Lexer struct {
	Curr  int
	Peek  int
	Char  byte
	Input string
}

func New(input string) *Lexer {
	lex := Lexer{Input: input}
	lex.advance()
	return &lex
}

func (l *Lexer) NextToken() token.Token {
	for unicode.IsSpace(rune(l.Char)) {
		l.advance()
	}
	var tok token.Token

	if l.Peek < len(l.Input) {
		if ds := string(l.Input[l.Curr]) + string(l.Input[l.Peek]); ds == "//" {
			return token.New(token.COMMENT, l.readComment())
		} else if tt, ok := token.DoubleTokenMap[ds]; ok {
			l.advance()
			l.advance()
			return token.New(tt, ds)
		}
	}

	switch l.Char {
	case 0:
		return token.New(token.EOF, "0")
	case '!':
		tok = token.New(token.EXCLAMATION, "!")
	case '@':
		tok = token.New(token.AT, "@")
	case '#':
		tok = token.New(token.HASHTAG, "#")
	case '$':
		tok = token.New(token.DOLLAR, "$")
	case '%':
		tok = token.New(token.PERCENT, "%")
	case '^':
		tok = token.New(token.CARET, "^")
	case '&':
		tok = token.New(token.AMPERSAND, "&")
	case '*':
		tok = token.New(token.ASTERISK, "*")
	case '(':
		tok = token.New(token.LPAREN, "(")
	case ')':
		tok = token.New(token.RPAREN, ")")
	case '-':
		tok = token.New(token.MINUS, "-")
	case '_':
		tok = token.New(token.UNDERSCORE, "_")
	case '+':
		tok = token.New(token.PLUS, "+")
	case '=':
		tok = token.New(token.ASSIGN, "=")
	case '[':
		tok = token.New(token.LBRACKET, "[")
	case ']':
		tok = token.New(token.RBRACKET, "]")
	case '{':
		tok = token.New(token.LBRACE, "{")
	case '}':
		tok = token.New(token.RBRACE, "}")
	case ';':
		tok = token.New(token.SEMICOLON, ";")
	case ':':
		tok = token.New(token.COLON, ":")
	case '\'':
		tok = token.New(token.CHAR, l.readChar())
	case '"':
		tok = token.New(token.STRING, l.readString())
	case ',':
		tok = token.New(token.COMMA, ",")
	case '.':
		tok = token.New(token.PERIOD, ".")
	case '<':
		tok = token.New(token.LESSTHAN, "<")
	case '>':
		tok = token.New(token.GREATERTHAN, ">")
	case '/':
		tok = token.New(token.SLASH, "/")
	case '?':
		tok = token.New(token.QUESTION, "?")
	case '\\':
		tok = token.New(token.BACKSLASH, "\\")
	case '|':
		tok = token.New(token.PIPE, "|")
	default:
		if unicode.IsLetter(rune(l.Char)) {
			ident := l.readIdentifier()
			if ttype, ok := token.IdentifierMap[ident]; ok {
				tok = token.New(ttype, ident)
			} else {
				tok = token.New(token.IDENTIFIER, ident)
			}
			return tok
		} else if unicode.IsNumber(rune(l.Char)) {
			ttype, lit := l.readNumber()
			return token.New(ttype, lit)
		} else {
			tok = token.New(token.ILLEGAL, "ILLEGAL")
		}
	}
	l.advance()

	return tok
}

func (l *Lexer) readIdentifier() string {
	pos := l.Curr
	for unicode.IsLetter(rune(l.Char)) || unicode.IsDigit(rune(l.Char)) || l.Char == '_' {
		l.advance()
	}
	return l.Input[pos:l.Curr]
}

func (l *Lexer) readNumber() (token.TokenType, string) {
	pos := l.Curr
	var is_float = false
	for unicode.IsDigit(rune(l.Char)) || l.Char == '.' {
		if l.Char == '.' {
			if is_float {
				l.advance()
				return token.ILLEGAL, "ILLEGAL"
			}
		}
		l.advance()
	}
	if is_float {
		return token.FLOAT, l.Input[pos:l.Curr]
	} else {
		return token.NUMBER, l.Input[pos:l.Curr]
	}
}

func (l *Lexer) readString() string {
	l.advance()
	pos := l.Curr
	for l.Char != '"' {
		l.advance()
	}
	return l.Input[pos:l.Curr]
}

func (l *Lexer) readChar() string {
	l.advance()
	pos := l.Curr
	for l.Char != '\'' {
		l.advance()
	}
	return l.Input[pos:l.Curr]
}

func (l *Lexer) readComment() string {
	pos := l.Curr
	for l.Char != '\n' && l.Char != '\r' {
		l.advance()
	}
	return l.Input[pos:l.Curr]
}

func (l *Lexer) advance() {
	if l.Peek >= len(l.Input) {
		l.Char = 0
	} else {
		l.Char = l.Input[l.Peek]
	}
	l.Curr = l.Peek
	l.Peek += 1
}

File: Clojure/Lexer/src/lexer/lexer.clj
(ns lexer.lexer
  (:require [lexer.token :refer [->Token singleTokenMap doubleTokenMap keywordMap]]
            [clojure.string :as str]))

(defrecord Lexer [input char curr peek])

(defn advance [lexer]
  (let [p (:peek lexer)
        ch (if (< p (count (:input lexer)))
             (nth (:input lexer) p)
             nil)]
    (->Lexer (:input lexer) ch p (inc p))))

(defn newLexer [input]
  (advance (->Lexer input nil 0 0)))

(defn isLetter [c]
  (when c (or (Character/isLetterOrDigit c) (= c \_))))

(defn isDigit [c]
  (when c (Character/isDigit c)))

(defn isWhitespace [c]
  (when c (<= (int c) 32)))

(defn readWhile [lexer pred]
  (loop [acc [] l lexer]
    (let [c (:char l)]
      (if (and c (pred c))
        (recur (conj acc c) (advance l))
        [(apply str acc) l]))))

(defn readIdentifier [lexer]
  (readWhile lexer isLetter))

(defn readNumber [lexer]
  (readWhile lexer #(or (isDigit %) (= % \.))))

(defn readString [lexer]
  (readWhile lexer #(not= % \")))

(defn readChar [lexer]
  (readWhile lexer #(not= % \')))

(defn readComment [lexer]
  (readWhile lexer #(not (contains? #{0, 10, 12, 13} (int %)))))

(defn nextToken [lexer]
  (let [{:keys [input curr char]} lexer
        ds (str/join "" (take 2 (drop curr input)))]
    (cond
      (isWhitespace char) (nextToken (advance lexer))

      ;; Early return for comments
      (= ds "//")
      (let [[literal new-lexer] (readComment lexer)]
        [(->Token :comment literal) (advance new-lexer)])

      ;; Double toks
      (contains? doubleTokenMap ds)
      (let [token-type (get doubleTokenMap ds)]
        [(->Token token-type ds) (advance (advance lexer))])

      ;; String literal
      (= char \")
      (let [[literal new-lexer] (readString (advance lexer))]
        [(->Token :string literal) (advance new-lexer)])

      ;; Char literal
      (= char \')
      (let [[literal new-lexer] (readChar (advance lexer))]
        [(->Token :char literal) (advance new-lexer)])

      ;; Single toks
      (contains? singleTokenMap char)
      (let [token-type (get singleTokenMap char)]
        [(->Token token-type (str char)) (advance lexer)])

      ;; a-zA-Z
      (isLetter char)
      (let [[ident new-lexer] (readIdentifier lexer)
            token-type (get keywordMap ident)]
        (if (nil? token-type)
          [(->Token :identifier ident) new-lexer]
          [(->Token token-type ident) new-lexer]))

      ;; 0-9
      (isDigit char)
      (let [[num new-lexer] (readNumber lexer)
            num-decimal (count (filter #{\.} num))
            token-type (case num-decimal 0 :number 1 :float :illegal)]
        [(->Token token-type num) new-lexer])

      ;; otherwise
      (nil? char) [(->Token :eof "eof") lexer]
      :else [(->Token :illegal (str char)) (advance lexer)])))

(defn parseWhile [lexer]
  (let [[tok new-lexer] (nextToken lexer)]
    (println tok)
    (cond
      (= (:type tok) :eof) :ok
      :else (recur new-lexer))))
